### PR TITLE
Remote cluster - API key security model - cluster privileges

### PR DESCRIPTION
**Pull request by jakelandis**
This commit adds a new `remote_cluster` to the role and API keys to enable expressing cluster permissions across remote clusters similar to the `remote_indices` block for RCS 2.0 (API key based cross cluster security).
```
"remote_cluster": [
        {
            "privileges": [
                "monitor_enrich"
            ],
            "clusters": [
                "my_remote*"
            ]
        }
    ]
```
The motivation for this change to enable ES|QL executed over remote clusters to allow or disallow remote enrichment. As such, the only privilege allowed here is "monitor_enrich" which controls if users can use ES|QL ENRICH keyword as part of their queries for the remote cluster. RemoteClusterSecurityEsqlIT has the relevant functional tests since the only way to exercise this permission. This change is quite large since this change touches the role and role descriptors used for API keys. That requires updates and/or testing for CRUD role, API keys, file based roles, get user privileges (which echos the role), usage stats, and the various views of the role : role descriptors, roles (base/limited by), role intersections, and (effective) roles. 

The data model design differs a bit from remote_indices in that it is a simpler data model which shares very little with the non-remote variant of cluster privileges. This allows the data to modeled only once via RemoteClusterPermissions and RemoteClusterPermissionGroup. RemoteClusterPermissions has N RemoteClusterPermissionGroup and is reused across the various vies of the roles. RemoteClusterPermissions is created early and is used as the primary contract with this new concept. Those data objects are self de/serializing and self toXcontent (we still use old school fromXContent and didn't want to mix and match old/new parsing strategies). Much like remote_indices, on transfer to the remote cluster, they will be converted into the non remote variant. 

Also note that in this PR, the monitor_enrich privilege is explicitly auto added to the backing API key set of permissions for CCS.  `CCS_CLUSTER_PRIVILEGE_NAMES = { "cross_cluster_search", "monitor_enrich" };` ideally, we would group monitor_enrich inside of cross_cluster_search privilege and avoid needing to explicitly add the permission. However, due to some low level details w.r.t. how the has_privilege is used, this is required. I will follow up to see if we can change how has_privilege is called via ES|QL to avoid the need to explicitly add monitor_enrich to the backing API key (infavor of only having cross_cluster_search that encompases the needed privilege). 

Also note, that today still need a local privilege of   "cluster": ["cross_cluster_search"], (or more granular  "cluster": ["monitor_enrich"] ) since ES|QL + ENRICH does not support remote only ENRICH. Also, ES|QL (with out enrich) over CCS requires the local privilege  "indices": [    {  "names": [""],       "privileges": ["read_cross_cluster"]    }  ] since the local action that coordinates the workflow needs to execute locally. 

Also, docs will come in a different PR.

### CODE REVIEW

File: x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/security/action/apikey/CrossClusterApiKeyRoleDescriptorBuilder.java [hunk 1]
```diff
@@ -18,14 +18,19 @@
...
 import java.util.List;
+import java.util.stream.Stream;
 
 import static org.elasticsearch.xcontent.ConstructingObjectParser.optionalConstructorArg;
 
 public class CrossClusterApiKeyRoleDescriptorBuilder {
 
-    public static final String[] CCS_CLUSTER_PRIVILEGE_NAMES = { "cross_cluster_search" };
+    // monitor_enrich is needed for ES|QL + ENRICH and https://github.com/elastic/elasticsearch/issues/106926 is related
+    public static final String[] CCS_CLUSTER_PRIVILEGE_NAMES = { "cross_cluster_search", "monitor_enrich" };
```
Comment (jakelandis) [inline x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/security/action/apikey/CrossClusterApiKeyRoleDescriptorBuilder.java:28]: I think i can remove the need to explicitly add this by changing how ES|QL calls has_privileges and will follow up in a seperate PR.

File: x-pack/plugin/security/qa/multi-cluster/src/javaRestTest/java/org/elasticsearch/xpack/remotecluster/RemoteClusterSecurityEsqlIT.java [hunk 2]
```diff
@@ -324,20 +330,23 @@ public void testCrossClusterQuery() throws Exception {
...
             | SORT emp_id ASC
             | LIMIT 2
-            | KEEP emp_id, department"""));
-        assertOK(response);
+            | KEEP emp_id, department""");
+        Response response = performRequestWithRemoteSearchUser(request);
+        assertRemoteOnlyResults(response);
+
+        // same as above but authenticate with API key
+        response = performRequestWithRemoteSearchUserViaAPIKey(request);
```
Comment (jakelandis) [inline x-pack/plugin/security/qa/multi-cluster/src/javaRestTest/java/org/elasticsearch/xpack/remotecluster/RemoteClusterSecurityEsqlIT.java:342]: this isn't related to the change here ... i just forgot to test this via API key in the prior PR.

**Review by slobodanadamovic — CHANGES_REQUESTED** [review]
The changes look pretty good. Nice job on testing!  I appreciate the javadoc comments. 

Most of the comments I left are optional, but as discussed via chat, we still need to address BWC concern when sending remote cluster permissions to a RCS 2.0 cluster on an older version. I think this should warrant a new BWC test. As an example, you can see how `RemoteClusterSecurityBwcRestIT` is [defined](https://github.com/elastic/elasticsearch/blob/e784706b4e9ddc5ba58ac45d031c2be8dc0a6499/x-pack/plugin/security/qa/multi-cluster/src/javaRestTest/java/org/elasticsearch/xpack/remotecluster/RemoteClusterSecurityBwcRestIT.java#L45) and [built](https://github.com/elastic/elasticsearch/blob/e784706b4e9ddc5ba58ac45d031c2be8dc0a6499/x-pack/plugin/security/qa/multi-cluster/build.gradle#L29-L41). Unfortunately, this test cannot be extended to test this scenario but it can serve as an example how to add a new BWC test.

**Comment by slobodanadamovic** [comment]
Do we want to track the usage of roles with `remote_cluster` privileges like we do for `remote_indices` ([here](https://github.com/elastic/elasticsearch/blob/fd6c2677953228d909ca05833e03405686298d3f/x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/authz/store/NativeRolesStore.java#L393-L396) and [here](https://github.com/elastic/elasticsearch/blob/fd6c2677953228d909ca05833e03405686298d3f/x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/authz/store/FileRolesStore.java#L156))?

**Comment by jakelandis** [comment]
>  we still need to address BWC concern when sending remote cluster permissions to a RCS 2.0 cluster on an older version. 

Good catch. I have updated the code to be remote version aware. 
* https://github.com/elastic/elasticsearch/pull/107493/commits/ed8fde7839a653b880d99d0d6cd66c6e6b6cdff4 -> plumbing needed to push the remote cluster version to check
* https://github.com/elastic/elasticsearch/pull/107493/commits/e01974a7e0c2c4aca57c6ae355b1cc8a8a1e3419 -> implement per version logic and unit test
* https://github.com/elastic/elasticsearch/pull/107493/commits/e1da2593f95c6d535c81c6e9cd0ca08e1fcba226 -> fix stuff i broke

> I think this should warrant a new BWC test. 

I don't think it does. I think a manual test is warranted, but this will likely change very infrequently, and the test would only be applicable one version back (due to when we GA RCS 2.0). There is not much nuance since it will work or it will not, due to we either send valid permissions or not, and the unit tests cover the logic for when we send permissions. There are some non-trivial costs in CPU, flakeyness, complexity, and time to write these BWC tests. In this case I think a quick spot check manually + the unit tests to ensure we send the write permission to the right version is sufficient.

**Comment by jakelandis** [comment]
> Do we want to track the usage of roles with remote_cluster privileges like we do for remote_indices ([here](https://github.com/elastic/elasticsearch/blob/fd6c2677953228d909ca05833e03405686298d3f/x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/authz/store/NativeRolesStore.java#L393-L396) and [here](https://github.com/elastic/elasticsearch/blob/fd6c2677953228d909ca05833e03405686298d3f/x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/authz/store/FileRolesStore.java#L156))?

Indeed.  I already got the filestore, but missed the native store. Thanks! -> https://github.com/elastic/elasticsearch/pull/107493/commits/cdae9b85f2e5f1717a343060d59beaf7c952bc51

File: x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/security/authz/RoleDescriptor.java [hunk 3]
```diff
@@ -700,12 +739,69 @@ private static RemoteIndicesPrivileges parseRemoteIndex(String roleName, XConten
...
+            XContentParser.Token token;
+            String currentFieldName = null;
+            while ((token = parser.nextToken()) != XContentParser.Token.END_OBJECT) {
+                if (token == XContentParser.Token.FIELD_NAME) {
+                    currentFieldName = parser.currentName();
+                } else if (Fields.PRIVILEGES.match(currentFieldName, parser.getDeprecationHandler())) {
+                    privileges = readStringArray(roleName, parser, false);
+                    if (privileges.length != 1
+                        || RemoteClusterPermissions.getSupportRemoteClusterPermissions()
+                            .contains(privileges[0].trim().toLowerCase(Locale.ROOT)) == false) {
```
[the remark this answers is not in the document]
Comment (jakelandis) [reply x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/security/authz/RoleDescriptor.java:770]: I think not trimming them is a low level bug. I will keep this here as i think that correctness > consistency here.

File: x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/support/SecuritySystemIndices.java [hunk 4]
```diff
@@ -295,6 +295,23 @@ private XContentBuilder getMainIndexMappings() {
                     }
                     builder.endObject();
 
+                    builder.startObject("remote_cluster");
```
[the remark this answers is not in the document]
Comment (jakelandis) [reply x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/support/SecuritySystemIndices.java:298]: good call, i missed that ! I will follow up on this one... but i think your changes for description will grandfather this one in (if not i will handle it in a follow up)

File: x-pack/plugin/core/src/test/java/org/elasticsearch/xpack/core/security/authz/permission/RemoteClusterPermissionsTests.java [hunk 5]
```diff
@@ -0,0 +1,168 @@
...
+
+    public void testPrivilegeNames() {
+        for (int i = 0; i < generateRandomGroups(true).size(); i++) {
+            String[] privileges = groupPrivileges.get(i);
+            String[] clusters = groupClusters.get(i);
+            for (String cluster : clusters) {
+                Arrays.sort(privileges);
+                String[] found = remoteClusterPermission.privilegeNames(cluster);
+                Arrays.sort(found);
+                assertArrayEquals(privileges, found);
```
[the remark this answers is not in the document]
Comment (jakelandis) [reply x-pack/plugin/core/src/test/java/org/elasticsearch/xpack/core/security/authz/permission/RemoteClusterPermissionsTests.java]: thanks for the pointer, but will leave for now.

File: x-pack/qa/rolling-upgrade/src/test/java/org/elasticsearch/upgrades/ApiKeyBackwardsCompatibilityIT.java [hunk 6]
```diff
@@ -425,7 +426,10 @@ private static RoleDescriptor randomRoleDescriptor(boolean includeRemoteIndices)
             generateRandomStringArray(5, randomIntBetween(2, 8), false, true),
             RoleDescriptorTests.randomRoleDescriptorMetadata(false),
             Map.of(),
-            includeRemoteIndices ? RoleDescriptorTests.randomRemoteIndicesPrivileges(1, 3, excludedPrivileges) : null,
+            includeRemoteDescriptors ? RoleDescriptorTests.randomRemoteIndicesPrivileges(1, 3, excludedPrivileges) : null,
+            includeRemoteDescriptors
+                ? RoleDescriptorTests.randomRemoteClusterPermissions(randomIntBetween(1, 3))
+                : RemoteClusterPermissions.NONE,
```
[the remark this answers is not in the document]
Comment (jakelandis) [reply x-pack/qa/rolling-upgrade/src/test/java/org/elasticsearch/upgrades/ApiKeyBackwardsCompatibilityIT.java:432]: see below.

**Comment by jakelandis** [comment]
all 👍 code review nits included in https://github.com/elastic/elasticsearch/pull/107493/commits/9b7ab458a08b8bd682ae49bce15133e11a40cc92

**Review by jeramysoucy — APPROVED** [review]
Kibana security owned change LGTM. I took a look through the rest of the changes as well.

File: x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/security/authz/RoleDescriptor.java (hunk 3 above)
[the remark this answers is not in the document]
Comment (slobodanadamovic) [reply x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/security/authz/RoleDescriptor.java:770]: I would be +1 on using `trim` when parsing privilege names. But here we are parsing them without trimming, which makes the check accept e.g. `monitor_enrich         ` which then fails validation later.

**Comment by slobodanadamovic** [comment]
> > I think this should warrant a new BWC test.
> 
> I don't think it does. I think a manual test is warranted, but this will likely change very infrequently, and the test would only be applicable one version back (due to when we GA RCS 2.0). There is not much nuance since it will work or it will not, due to we either send valid permissions or not, and the unit tests cover the logic for when we send permissions. There are some non-trivial costs in CPU, flakeyness, complexity, and time to write these BWC tests. In this case I think a quick spot check manually + the unit tests to ensure we send the write permission to the right version is sufficient.

Good point. 👍

**Review by slobodanadamovic — APPROVED** [review]
LGTM 👍

File: x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/security/authz/permission/RemoteClusterPermissions.java [hunk 7]
```diff
@@ -0,0 +1,221 @@
...
+import java.util.HashSet;
+import java.util.List;
+import java.util.Locale;
+import java.util.Map;
+import java.util.Objects;
+import java.util.Set;
+import java.util.stream.Collectors;
+
+/**
+ * Represents a group of permissions for a remote cluster. This is intended to be the model for both the {@link RoleDescriptor}
```
[the remark this answers is not in the document]
Comment (jakelandis) [reply x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/security/authz/permission/RemoteClusterPermissions.java]: thanks ! clarified with https://github.com/elastic/elasticsearch/pull/107493/commits/40a52567f0ca20f95b01ee66f67521ecc5ae1a20

**Pull request merged**